extension/framework/goplugin/runner/runner.go

```diff
@@ -19,208 +19,237 @@ import (
 	"github.com/cloudwan/gohan/extension/goext"
 	"github.com/cloudwan/gohan/extension/goplugin"
-	pkgLog "github.com/cloudwan/gohan/log"
+	gohan_logger "github.com/cloudwan/gohan/log"
 	"github.com/cloudwan/gohan/schema"
 	"github.com/cloudwan/gohan/server/middleware"
 	"github.com/cloudwan/gohan/sync/noop"
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
+	"github.com/twinj/uuid"
 )
 
 const (
-	testDBFile = "test.db"
+	dbBaseFileName = "test.db"
 )
 
-var log = pkgLog.NewLogger()
+var log = gohan_logger.NewLogger()
 
-// GoTestRunner is a test runner for go (plugin) extensions
-type GoTestRunner struct {
+// TestRunner is a test runner for go extensions
+type TestRunner struct {
 	pluginFileNames []string
-	printAllLogs    bool
-	testFilter      *regexp.Regexp
-	workers         int
+	verboseLogs     bool
+	fileNameFilter  *regexp.Regexp
+	workerCount     int
 }
 
-func dbConnString(name string) string {
-	return fmt.Sprintf("file:%s?mode=memory&cache=shared", name)
+// NewTestRunner allocates a new TestRunner
+func NewTestRunner(pluginFileNames []string, printAllLogs bool, testFilter string, workers int) *TestRunner {
+	return &TestRunner{
+		pluginFileNames: pluginFileNames,
+		verboseLogs:     printAllLogs,
+		fileNameFilter:  regexp.MustCompile(testFilter),
+		workerCount:     workers,
+	}
 }
 
-func dbConnect(connString string) (db.DB, error) {
-	conn, err := db.ConnectDB("sqlite3", connString, db.DefaultMaxOpenConn, options.Default())
-	if err != nil {
-		return nil, err
+// Run runs go test runner
+func (testRunner *TestRunner) Run() error {
+	// configure reporter
+	gomega.RegisterFailHandler(ginkgo.Fail)
+	t := &testing.T{}
+	reporter := NewReporter()
+
+	// run suites
+	var err error
+
+	for _, pluginFileName := range testRunner.pluginFileNames {
+		if !testRunner.fileNameFilter.MatchString(pluginFileName) {
+			continue
+		}
+
+		if err = testRunner.runSingle(t, reporter, pluginFileName); err != nil {
+			break
+		}
 	}
-	return conn, nil
+
+	reporter.Report()
+	return err
 }
 
-// NewGoTestRunner allocates a new GoTestRunner
-func NewGoTestRunner(pluginFileNames []string, printAllLogs bool, testFilter string, workers int) *GoTestRunner {
-	return &GoTestRunner{
-		pluginFileNames: pluginFileNames,
-		printAllLogs:    printAllLogs,
-		testFilter:      regexp.MustCompile(testFilter),
-		workers:         workers,
+func readSchemas(p *plugin.Plugin) ([]string, error) {
+	fnRaw, err := p.Lookup("Schemas")
+
+	if err != nil {
+		return nil, fmt.Errorf("missing 'Schemas' export: %s", err)
+	}
+
+	fn, ok := fnRaw.(func() []string)
+
+	if !ok {
+		return nil, fmt.Errorf("invalid signature of 'Schemas' export")
 	}
+
+	return fn(), nil
 }
 
-// GoTestSuite is a test suite state for a go (plugin) extension runner
-type GoTestSuite struct {
-	plugin  *plugin.Plugin
-	db      db.DB
-	env     *goplugin.Environment
-	path    string
-	manager *schema.Manager
+func readBinaries(p *plugin.Plugin) ([]string, error) {
```

Comment:
currently the test suites are bloated with a long list of schemas followed by a long list of binaries. is it possible to load the binaries specified in schemas instead providing a separate list?

Reply:
It is not possible in all cases. There are schemas which contain more than one plugin and therefore the test would load all of the plugins from a specific schema. This is not the desired behavior. Unit-tests can focus on testing only a single plugin and for that case it shouldn't be needed to create a special schema just to load this single plugin.